### ISSUE TITLE
Tweak query radius

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/prox#177]
Follow-up to #49.

From the [criteria doc](https://docs.google.com/document/d/101jsPjZZ-a_MgZn1UaAyJouufkxzJqG07IrI_0USj1M/edit#), we show UX the following radius and get feedback as to what is most appropriate:

> 0.5 hour driving away
1 hour driving away
1.5 hour driving away
2 hour driving away

**Comment by mcomella** [comment]
@antlam, how do you want to see the data in order to verify the radius? Do you need to see it on device? Can we give you json? Will a count of the number of places it includes suffice (at least to eliminate the way-too-large radii)?

In order to get the data for UX, we'd need to:
* Query for all places within the distances above – this can only be done from the server (or directly with yelp's APIs)
  * CC @jhugman for awareness
* Filter places based on the category (see #178) & rating (see #179) criteria UX has already specified (post #49, the logic is only available on the client)

When implementing finally, we'll need to:
* Make sure all the data around the hotel for that space is pre-cached
* Make sure the client -> server query for the users' current location uses the radius (unless there are too many results, in which case maybe we should sort results by distance and stop when the number gets large enough – we might miss some big places though)
* We might want to consider: making multiple queries at different radius to gauge how valuable the results are at each level

fwiw:
* [0.5hr driving][0.5hr] ~= [27km](https://www.mapdevelopers.com/distance_finder.php?&points=%5B%5B19.92216463414154%2C-155.885009765625%5D%2C%5B19.73348610091148%2C-156.03933334350586%5D%5D)
* [1hr driving][1hr] ~= [48km](https://www.mapdevelopers.com/distance_finder.php?&points=%5B%5B19.923455737182163%2C-155.885009765625%5D%2C%5B19.495204521892845%2C-155.9183120727539%5D%5D)
* [1.5hr driving][1.5hr] ~= [73km](https://www.mapdevelopers.com/distance_finder.php?&points=%5B%5B19.923132962410595%2C-155.88638305664062%5D%2C%5B19.271091277164587%2C-155.87574005126953%5D%5D)
* [2hr driving][2hr] ~= [115km (basically the whole island)](https://www.mapdevelopers.com/distance_finder.php?&points=%5B%5B19.92474682967762%2C-155.88638305664062%5D%2C%5B19.363429595526743%2C-154.9683380126953%5D%5D)

[0.5hr]: https://www.google.com/maps/dir/Hilton+Waikoloa+Village,+Waikoloa+Beach+Drive,+Waikoloa+Village,+HI/Kona+International+Airport,+73-200+Kupipi+Street,+Kailua-Kona,+HI+96740/@19.7084239,-156.0466751,12.88z/data=!4m13!4m12!1m5!1m1!1s0x7954778ec6a99dc5:0x65cec723a9eaa336!2m2!1d-155.887498!2d19.924029!1m5!1m1!1s0x79540d7d01ffe7d3:0xab18bf0cbb3f5937!2m2!1d-156.0429246!2d19.736916
[1hr]: https://www.google.com/maps/dir/Hilton+Waikoloa+Village,+Waikoloa+Beach+Drive,+Waikoloa+Village,+HI/Captain+Cook,+HI/@19.7103097,-156.0937406,11z/data=!3m1!4b1!4m13!4m12!1m5!1m1!1s0x7954778ec6a99dc5:0x65cec723a9eaa336!2m2!1d-155.887498!2d19.924029!1m5!1m1!1s0x795406c8487722e1:0x1296f1d0ab999a67!2m2!1d-155.9216667!2d19.4969444
[1.5hr]: https://www.google.com/maps/dir/Hilton+Waikoloa+Village,+Waikoloa+Beach+Drive,+Waikoloa+Village,+HI/Kipahoehoe+Natural+Area+Reserve/@19.5538009,-155.9831907,11z/data=!4m13!4m12!1m5!1m1!1s0x7954778ec6a99dc5:0x65cec723a9eaa336!2m2!1d-155.887498!2d19.924029!1m5!1m1!1s0x0:0x6996a77bfd6ac09e!2m2!1d-155.8382291!2d19.2727133
[2hr]: https://www.google.com/maps/dir/Kalapana,+HI/Hilton+Waikoloa+Village,+69-425+Waikoloa+Beach+Dr,+Waikoloa+Village,+HI+96738/@19.638207,-155.9744475,9z/data=!3m1!4b1!4m13!4m12!1m5!1m1!1s0x7952294ad19b6f7f:0x9701b456cc979171!2m2!1d-154.97951!2d19.354803!1m5!1m1!1s0x7954778ec6a99dc5:0x65cec723a9eaa336!2m2!1d-155.887498!2d19.924029

**Comment by mcomella** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [As a user I want to see the most useful places when I open the app so that I can make better decisions about my next step](https://trello.com/c/IBEHIXNb/149-as-a-user-i-want-to-see-the-most-useful-places-when-i-open-the-app-so-that-i-can-make-better-decisions-about-my-next-step)

**Comment by antlam** [comment]
@mcomella Thanks for the deets!

To finalize the radius we want to preset, I think we need to see the results on an actual device. 

Given that, let's start with 0.5, 1, and 1.5 to start. 2 hr was really to see what the limits were and we might not actually need it right now.

**Comment by mcomella** [comment]
@jhugman How can we get the server to return all places within a given radius for UX to see? Would we need to filter by categories/rating on the server or can we wait until the data arrives on the client?

**Comment by jhugman** [comment]
@mcomella @antlam getting the server to return all places within a given radius – do you just want some more JSON to review the criteria?

For 'here's some JSON for UX' type tasks, `samples.crawl-yelp` will generate lots of JSON for you.

We can certainly filter by categories/rating on the server. For batteries' sakes I'd prefer to do the bulk of the filtering there, then fine tune it on the client.

Lastly: for quick'n'dirty hitting the yelp search API: there is a 40km limit, so just under an hour's drive. :(

**Comment by jhugman** [comment]
@mcomella We should be in a better state to start tweaking things now that we have [RemoteConfig available][1].

@antlam I've attached a new batch of JSON for you ;) at https://gist.github.com/jhugman/23424345cb79b091fc335ab7e5eaf7b9

  [1]: https://github.com/mozilla-mobile/prox/pull/173

**Comment by mcomella** [comment]
Spoke with @jhugman on vidyo:
* Yelp caps the search radius at 40km
  *  Getting the results for a larger area is a half a day's work (James expects half a day for him to do it, I assume more for other people)
  * (As mentioned above) James made this 40km query around the hotel and [put the data in json](https://gist.github.com/jhugman/23424345cb79b091fc335ab7e5eaf7b9) – due to communication errors, it is unclear if this was using [the latest category criteria](https://github.com/mozilla-mobile/prox/issues/49#issuecomment-260531337). Instead, it may have been whitelisting [these categories](https://github.com/mozilla-mobile/prox-server/pull/43/files).
    * I think it'd be a small amount of additional work to get this data into the app directly
* Concern: the new category criteria (blacklisting, rather than whitelisting) means we'd have to get *all* the places on the island then filter them by category (as opposed to requesting a subset of places by specifying categories)
  * This is non-trivial because yelp caps its results at 1k items even if Yelp is aware of a larger number
    * In the 40km radius from the hotel, there are 2343 places. Roughly [half the search radius is over water](https://www.mapdevelopers.com/draw-circle-tool.php?circles=%5B%5B40000%2C19.9237934%2C-155.8873798%2C%22%23AAAAAA%22%2C%22%23000000%22%2C0.4%5D%5D) so it'd probably get worse inland (though there are parks)

@antlam, let's talk today about solutions!

**Comment by mcomella** [comment]
Spoke with @antlam:
* Primary goal is to see and compare a listing of places at various radius
  * Would prefer to see on device but will take json if not
* Data should be filtered by [category criteria](https://github.com/mozilla-mobile/prox/issues/49#issuecomment-260531337) and [rating criteria](https://github.com/mozilla-mobile/prox/blob/master/Prox/Prox/Utilities/PlaceUtilities.swift#L49)
  * The category criteria makes these queries challenging but I confirmed with antlam that this is what we need to do

Since it's a lot easier to get json, I will get the 0.5hr radius for antlam.

However, 1hr & 1.5hr radius are hard because we have to make multiple queries – @jhugman, can you take care of this?

To be explicit, the category criteria makes this hard: we "hide places whose categories (all of them) we said to hide" meaning we need to retrieve *all* places from yelp and filter them on the server, which prevents us from reducing the places we receive from yelp by specifying explicit categories. This means the max 40km radius will generally exceed 1000 places and, if we don't reduce the circle, we'll be missing places.

**Comment by mcomella** [comment]
Here is [the json for 27km around Kona](https://gist.github.com/mcomella/1dd09d67a16cec28b40bba358d5b708c) and [the code I used](https://github.com/mcomella/location-api-comparison/blob/2a72c53f3a605c1dfebe316c4b49ee4cf6f7f3bc/yelp_hawaii.py).

Caveats:
* There were 1028 original place results but since Yelp only lets us get 1k, we lost 28.
* I followed the rating logic in the doc but noticed the way I had originally written it was not what was intended so the logic may not be entirely right – we should double-check with UX for what they want

**Comment by antlam** [comment]
@mcomella questions!

1. Can I see the 28 that we lost? I'm worried there could be some "hole in the wall" places in there.
2. Not sure what you mean here... so is this list following the "Filtering Criteria" doc?

**Comment by mcomella** [comment]
> Can I see the 28 that we lost? I'm worried there could be some "hole in the wall" places in there.

Unfortunately, since yelp limits us to the 1k results it's impossible to know which they were without some complex querying (i.e. query the circle around our circle, find the lat/long that fit in the original geocircle, add them to the list), which I don't think is worth the time (I'd guess a full days work? Unless @jhugman can query the whole island so we don't have to do special querying).

fwiw, the returned values were in yelp's "best_match" mode – the 28 places we lost were the 28 worst places (based on distance, rating, review count, etc.); Yelp broke when I tried to sort by absolute distance.

> Not sure what you mean here... so is this list following the "Filtering Criteria" doc?

The ratings are following the filtering criteria doc but I think the logic is contradictory. Specifically:
```
If (rating = 3) && (review_count > 3) THEN hide
If (review_count <=3) THEN hide
EXCEPT when rating = 5
```
Based on line 1, if rating is 3 and review_count is 2, I assume we'd want to show the place. However, this would get caught in line 2: review_count <= 3 so we'd hide it. The corrected logic is:
```
rating == 3 && review_count > 3 then HIDE
if review_count <=3 and rating != 5 and rating != 3 then HIDE
```
But it seems weird to special case 3 & 5: I'm currently working with @vazquez to address this.

**Comment by mcomella** [comment]
Let's discuss ratings in #179.

**Comment by mcomella** [comment]
Here's [the updated json](https://gist.github.com/mcomella/391c5bda310909f243c0457969efb2c3) with the ratings criteria I discussed with Ricardo.

**Comment by mcomella** [comment]
Spoke with @liuche. Here is an overview of where we're at:
* My understanding is that @antlam is trying to see *all* the places within 0.5hr, 1hr, and 1.5hr driving times (matching our existing criteria) so that he can further tweak our query parameters (e.g. category, radius, and rating)
  * We have json for 0.5hr driving: I [found 626 places in 27km](https://gist.github.com/mcomella/391c5bda310909f243c0457969efb2c3) and @jhugman has done his own search (link?).
    * Note: @liuche pointed out a logic simplification so we can get those 28 places I missed
  * I will get 1hr and 1.5hr, if needed
  * Technical note: see [driving times to distances](https://github.com/mozilla-mobile/prox/issues/177#issuecomment-260807392)

**Comment by liuche** [comment]
@antlam semi-related to getting a json dump so you can fine-tune the filtering criteria, I have a question about how granular you want the actual results the server returns are. 

Leaving this link here because it's related: https://github.com/mozilla-mobile/prox-server/pull/51#issuecomment-261404102

**Comment by antlam** [comment]
@liuche I'll respond in there.

**Comment by mcomella** [comment]
**Final results**:
* [0.5hr driving][0.5hr json] (27km; 638 places)
* 1hr driving (48km)
  * [Subqueries][1hr sub]
* 1.5hr driving (73km)

I will update this post as the results come in.

[0.5hr json]: https://gist.github.com/mcomella/89e4673fb2a19b15d80d35aa0eafd4c1
[1hr json]: a
[1hr sub]: https://www.mapdevelopers.com/draw-circle-tool.php?circles=%5B%5B48000%2C19.9237934%2C-155.8873798%2C%22%23AAAAAA%22%2C%22%23000000%22%2C0.4%5D%2C%5B40000%2C19.9186289%2C-155.8104755%2C%22%23AD0000%22%2C%22%23000000%22%2C0.4%5D%2C%5B40000%2C19.8385582%2C-155.8489277%2C%22%230EA2AD%22%2C%22%23000000%22%2C0.4%5D%5D
[1.5hr json]: a
[1.5hr sub]: a

**Comment by mcomella** [comment]
If anyone needs the code I'm using to generate these lists, its [in the "prox-places" branch of my location-api-comparison repo](https://github.com/mcomella/location-api-comparison/tree/prox-places).

**Comment by mcomella** [comment]
@vazquez has started using similar lists for categories so I added them here: https://github.com/mozilla-mobile/prox/issues/178#issuecomment-262387903

**Comment by mcomella** [comment]
@antlam Is this still something up in the air or did you want to settle on 27km? We can open a new bug (or update it in firebase remote config) if it's problematic.

**Comment by mcomella** [comment]
@antlam said we'll probably finalize in the next few days.

**Comment by mcomella** [comment]
Spoke with @antlam, we're going to settle on 1km. I've updated Firebase and we'll merge this with the defaults in #250.

**Comment by antlam** [comment]
Additional context (mostly so I don't forget):

- 1 km is quite a bit smaller than the 0.5 hour drive we started at.
- 1 km gets us a more controllable/sensible number of results, given our current Yelp category filtering + Ratings filtering + Review count filtering
- This means there is less likely to be "driving directions", but that's OK (it wasn't something we designed the app for in the Design Sprint either).
- We'll keep a close eye and tweak this as necessary, but 1 km is what we're settling on for now so we can get things moving.
- Figuring out how to prioritize well know/interesting places outside of the immediate 1km radius is a priority next step
